library/common-library/templates/_security-context.tpl

```diff
@@ -0,0 +1,22 @@
+{{- /* Defines the container securityContext context */ -}}
+{{- define "common.securityContext.container" -}}
+{{- $global := index .Values "global" | default dict -}}
+
+{{- if .Values.containerSecurityContext -}}
```

Comment:
How can we provide default in charts without discarding alway the global one?

Example:
 - http://github.com/newrelic/helm-charts/blob/ecf47c05c7aee6c5b04500a57c643109b36fce7a/charts/nri-metadata-injection/values.yaml#L58-L58
 - http://github.com/newrelic/helm-charts/blob/ecf47c05c7aee6c5b04500a57c643109b36fce7a/charts/nri-metadata-injection/templates/admission-webhooks/job-patch/job-patchWebhook.yaml#L47-L47`

Even with the merge I imagine that with default we would always overwrite globals

Reply:
This is an interesting topic. I see three solutions, and I like neither of them:

1. Provide the chart "defaults" in the template rather than in the values. So Global values will override the template defaults, and local values will override globals.
2. Make global override local for this particular case
3. Add a boolean "helper" somewhere, like `global.securityContext.override` (defaulting to `false`) that will cause the global SC to override the local one when `true`.

Of all this I think the one I dislike less is 1️⃣ .

Reply:
I agree, I do not like 3 since it is not easy to be applied to other similar cases in the future.

I really dislike 2 since it would be absolutely counterintuitive and difficult to document.

I would go for the 1st as well

Reply:
@newrelic/coreint Any last thought on this?
I think we can go for 1️⃣ tomorrow if no one has a better approach.

Reply:
i can think of some but non less ugly than the 1

Reply:
I think the helper returning defaults is needed for booleans (like `hostNetwork`) as we cannot tell if the helper is returning empty or false, but in this case I think it might be more complex. We can simply do:

```yaml
{{- if include "common.securityContext.pod" . }}
  {{- include "common.securityContext.pod" . | toYaml | nindent 2 }}
{{- else }}
  runAsUser: 1000
  runAsGroup: 3000
  fsGroup: 2000
{{- end }}
```

Which is functionally the same as the `common.securityContext.podDefaults`, but might be more readable. WDYT?

Reply:
I think the helper would be useful because of ugly things like the discussion we have to raise again:
- [newrelic-infrastructure-v3] Change hostNetwork depending if their kind is DaemonSet or Deployment [#698 ](https://github.com/newrelic/helm-charts/pull/698)
- newrelic-infrastructure-v3: default to recreate strategy for controlPlane Deployment [#697](https://github.com/newrelic/helm-charts/pull/697)

If we need something like "change defaults depending on something external" is better to have a fallback helper that can be overridable and it can allow me to change defaults in case of other flags changing (like `privileged` which is the next PR I would have to do)

Having so many moving parts is the reason I prefer the 3️⃣. Because changing defaults are evil but we are doing it in all the Helm charts because of UX.

Reply:
The [latest commit](https://github.com/newrelic/helm-charts/pull/725/commits/0f4c594ecf9d3fcde4d859e4c41aedbd1fa4ca87) adds the ability to set defaults using a helper function because it is easier to revert it after finishing the discussion than do it later and have to wait for approvals.

Give me your thoughts, please.